applications/FluidDynamicsHydraulicsApplication/custom_utilities/hydraulic_fluid_auxiliary_utilities.cpp

```diff
@@ -0,0 +1,322 @@
+        if (r_cond.Is(rSkinFlag))
+        {
+            std::vector<ModelPart::IndexType> elem_nodes_id;
+            auto &r_geom = r_cond.GetGeometry();
+
+            // Fill the points array and distances vector
+            Vector aux_distances(3);
+            for (IndexType i_nodes = 0; i_nodes < r_geom.PointsNumber(); i_nodes++)
+            {
+                aux_points.push_back(r_geom(i_nodes));
+                aux_distances[i_nodes] = r_geom[i_nodes].GetValue(rDistanceVariable);
+                // TODO: It should be posible to have an historical distance variable.
+                // aux_distances[i_nodes] =distance_getter;
+            }
+            // Calculate the water area (wetted area) of the cut conditions
+            if (IsSplit(aux_distances))
+            {
+                auto p_splitting_util = Kratos::make_unique<DivideTriangle3D3<NodeType>>(r_geom, aux_distances);
+                p_splitting_util->GenerateDivision();
+                const auto& r_neg_subdivisions = p_splitting_util->GetNegativeSubdivisions();
+                for (const auto& rp_neg_subdivision : r_neg_subdivisions) {
+                    wetted_area += rp_neg_subdivision->Area();
+                }
+            }
+            // Calculate the water area (wetted area)
+            else if (IsNegative(aux_distances))
+            {
+                wetted_area += r_geom.Area();
+            }
+            aux_points.clear();
+        }
+    }
+    return wetted_area;
+}
+
+bool HydraulicFluidAuxiliaryUtilities::IsSplit(const Vector &rConditionDistancesVector)
+{
+    std::size_t n_pos(0);
+    std::size_t n_neg(0);
+    const std::size_t pts_number = rConditionDistancesVector.size();
+    for (std::size_t i_node = 0; i_node < pts_number; ++i_node)
+    {
+        if (rConditionDistancesVector[i_node] > 0.0)
+            n_pos++;
+        else
+            n_neg++;
+    }
+    return (n_pos > 0 && n_neg > 0) ? true : false;
+}
+
+bool HydraulicFluidAuxiliaryUtilities::IsNegative(const Vector &rConditionDistancesVector)
+{
+    std::size_t n_neg(0);
+    const std::size_t pts_number = rConditionDistancesVector.size();
+    for (std::size_t i_node = 0; i_node < pts_number; ++i_node)
+    {
+        if (rConditionDistancesVector[i_node] < 0.0)
+            n_neg++;
+    }
+    return n_neg == pts_number;
+}
+
+double HydraulicFluidAuxiliaryUtilities::InitialWaterDepth(ModelPart &rModelPart)
+{
+
+   //Determine the initial estimate for water depth by considering the average of the maximum and minimum coordinates.
+    KRATOS_WARNING("HydraulicFluidAuxiliaryUtilities") << "The water depth is assumed to be in the Z direction" << std::endl;
+    double max_value = -1e9;
+    double min_value = 1e9;
+    for (auto &r_node : rModelPart.Nodes())
+    {
+        if (r_node.Z() > max_value)
+        {
+            max_value = r_node.Z();
+        }
+        if (r_node.Z() < min_value)
+        {
+            min_value = r_node.Z();
+        }
+    }
```

Comment:
We can do this in parallel with a combined max - min reduction.

Reply:
working on